unit_tests/log/test_main_chkpt_info.cpp

```diff
@@ -0,0 +1,960 @@
+{
+  REQUIRE (td1->head_lsa == td2->head_lsa);
+  REQUIRE (td1->tail_lsa == td2->tail_lsa);
+  REQUIRE (td1->undo_nxlsa == td2->undo_nxlsa);
+  REQUIRE (td1->posp_nxlsa == td2->posp_nxlsa);
+  REQUIRE (td1->savept_lsa == td2->savept_lsa);
+  REQUIRE (td1->tail_topresult_lsa == td2->tail_topresult_lsa);
+  REQUIRE (td1->rcv.sysop_start_postpone_lsa == td2->rcv.sysop_start_postpone_lsa);
+  REQUIRE (td1->rcv.atomic_sysop_start_lsa == td2->rcv.atomic_sysop_start_lsa);
+}
+
+void
+full_compare_tdes_active (log_tdes *td1, const log_tdes *td2)
+{
+  REQUIRE (td1->head_lsa == td2->head_lsa);
+  REQUIRE (td1->tail_lsa == td2->undo_nxlsa);
+  REQUIRE (td1->tail_lsa == td2->tail_lsa);
+  REQUIRE (td1->posp_nxlsa == td2->posp_nxlsa);
+  REQUIRE (td1->savept_lsa == td2->savept_lsa);
+  REQUIRE (td1->tail_topresult_lsa == td2->tail_topresult_lsa);
+  REQUIRE (td1->rcv.sysop_start_postpone_lsa == td2->rcv.sysop_start_postpone_lsa);
+  REQUIRE (td1->rcv.atomic_sysop_start_lsa == td2->rcv.atomic_sysop_start_lsa);
+}
+
+void
+check_recovery (checkpoint_info obj)
+{
+  std::map<TRANID, log_tdes *>::iterator itr;
+  for (itr = tran_map.begin(); itr != tran_map.end(); ++itr)
+    {
+      REQUIRE (itr->first != NULL_TRANID);
+    }
+
+  for (int i = 1; i < log_Gl.trantable.num_total_indices; i++)
+    {
+      LOG_TDES *tdes = log_Gl.trantable.all_tdes[i];
+
+      if (log_Gl.trantable.all_tdes[i]->trid == NULL_TRANID)
+	{
+	  continue;
+	}
+
+      std::map<TRANID, log_tdes *>::const_iterator tdes_after = tran_map.find (tdes->trid);
+      if (tdes->trid != NULL_TRANID && tdes->tail_lsa == NULL_LSA)
+	{
+	  REQUIRE (tdes_after == tran_map.end());
+	  continue;
+	}
+      REQUIRE (tdes_after != tran_map.end());
+      if (tdes->state == TRAN_ACTIVE || tdes->state == TRAN_UNACTIVE_ABORTED)
+	{
+	  REQUIRE (tdes_after->second->state == TRAN_UNACTIVE_UNILATERALLY_ABORTED);
+
+	  if (tdes->state == TRAN_UNACTIVE_ABORTED)
+	    {
+	      full_compare_tdes_active (tdes, tdes_after->second);
+	    }
+	  else
+	    {
+	      full_compare_tdes (tdes, tdes_after->second);
+	    }
+	  continue;
+	}
+      REQUIRE (tdes_after->second->state == tdes->state);
+      full_compare_tdes (tdes, tdes_after->second);
+    }
+
+  for (itr = systb_System_tdes.begin(); itr != systb_System_tdes.end(); ++itr)
+    {
+      std::map<TRANID, log_tdes *>::const_iterator tdes_after = tran_map.find (itr->first);
+      // check topops/rcv only
+
+      if (tdes_after == tran_map.end())
+	{
+	  continue;
+	}
+
+      if (itr->second->topops.last == -1 || itr->second->rcv.sysop_start_postpone_lsa == NULL_LSA
+	  || itr->second->rcv.atomic_sysop_start_lsa == NULL_LSA)
+	{
+	  REQUIRE (tdes_after->second->topops.last == 0);
+	}
+      else
+	{
+	  REQUIRE (itr->second->rcv.sysop_start_postpone_lsa == tdes_after->second->rcv.sysop_start_postpone_lsa);
+	  REQUIRE (itr->second->rcv.atomic_sysop_start_lsa == tdes_after->second->rcv.atomic_sysop_start_lsa);
+	}
```

Comment:
Only sysops with not null sysop_start_postpone_lsa or not null atomic_sysop_start_lsa are loaded/recovered. See checkpoint_info::load_checkpoint_topop.

```suggestion

      if (itr->second->rcv.sysop_start_postpone_lsa == NULL_LSA && itr->second->rcv.sysop_start_postpone_lsa == NULL_LSA)
        {
          REQUIRE (tdes_after == tran_map.end ());
        }
      else
        {
          REQUIRE (tdes_after != tran_map.end ());
          REQUIRE (itr->second->rcv.sysop_start_postpone_lsa == tdes_after->second->rcv.sysop_start_postpone_lsa);
	  REQUIRE (itr->second->rcv.atomic_sysop_start_lsa == tdes_after->second->rcv.atomic_sysop_start_lsa);
	  REQUIRE (tdes_after->second->topops.last == 0);
        }
```